internal/p2p/channel.go

```diff
@@ -214,3 +219,135 @@ func MergedChannelIterator(ctx context.Context, chs ...Channel) *ChannelIterator
 
 	return iter
 }
+
+////////////////////////////////////////////////////////////////////////
+
+type libp2pChannelImpl struct {
+	chDesc  *ChannelDescriptor
+	ps      *pubsub.PubSub
+	h       host.Host
+	topic   *pubsub.Topic
+	chainID string
+	wrapper Wrapper
+}
+
+func NewLibP2PChannel(chainID string, chDesc *ChannelDescriptor, ps *pubsub.PubSub, h host.Host) (Channel, error) {
+	ch := &libp2pChannelImpl{chDesc: chDesc, ps: ps, h: h, chainID: chainID}
+	topic, err := ps.Join(ch.canonicalizedTopicName())
+	if err != nil {
+		return nil, err
+	}
+	ch.topic = topic
+
+	if w, ok := chDesc.MessageType.(Wrapper); ok {
+		ch.wrapper = w
+	}
+
+	// TODO(tychoish) register handlers for
+	// request/response patterns
+
+	return ch, nil
+}
+
+func (ch *libp2pChannelImpl) String() string {
+	return fmt.Sprintf("Channel<%s>", ch.canonicalizedTopicName())
+}
+
+func (ch *libp2pChannelImpl) canonicalizedTopicName() string {
+	return strings.Join([]string{ch.chainID, ch.chDesc.Name, fmt.Sprint(ch.chDesc.ID)}, ".")
+}
+
+func (ch *libp2pChannelImpl) Receive(ctx context.Context) *ChannelIterator {
+	iter := &ChannelIterator{
+		pipe: make(chan Envelope), // unbuffered
+	}
+
+	sub, err := ch.topic.Subscribe()
+	if err != nil {
+		return nil
+	}
+
+	go func() {
+		defer close(iter.pipe)
+		for {
+			msg, err := sub.Next(ctx)
+			if err != nil {
+				return
+			}
+
+			payload := proto.Clone(ch.chDesc.MessageType)
+			if err := proto.Unmarshal(msg.Data, payload); err != nil {
+				// TODO: add error reporting abilities
+				// so we don't just miss these errors.
+				return
+			}
+			if wrapper, ok := payload.(Wrapper); ok {
+				if payload, err = wrapper.Unwrap(); err != nil {
+					return
+				}
+			}
+
+			select {
+			case <-ctx.Done():
+				return
+			case iter.pipe <- Envelope{
+				From:      types.NodeID(msg.From),
+				Message:   payload,
+				ChannelID: ch.chDesc.ID,
+			}:
+			}
+
+		}
+	}()
+
+	return nil
+}
+
+func (ch *libp2pChannelImpl) Send(ctx context.Context, e Envelope) error {
+	if ch.wrapper != nil {
+		msg := proto.Clone(ch.wrapper)
+		if err := msg.(Wrapper).Wrap(e.Message); err != nil {
```

Comment:
ooogh this API (no action item here, just noting it)

Reply:
yeah, we might be able to get rid of it, but this seems to be the only salient thing we're doing in sendMessage at the moment, so it seems safe to keep around.